src/syscall/syscall_libc.go

```diff
@@ -48,18 +54,36 @@ func Getpid() (pid int) {
 }
 
 func Getenv(key string) (value string, found bool) {
-	return "", false // TODO
+	data := append([]byte(key), 0)
+	raw := libc_getenv(&data[0])
+	if raw == nil {
+		return "", false
+	}
+
+	ptr := uintptr(unsafe.Pointer(raw))
+	for size := uintptr(0); ; size++ {
+		v := *(*byte)(unsafe.Pointer(ptr))
+		if v == 0 {
+			src := *(*[]byte)(unsafe.Pointer(&sliceHeader{buf: raw, len: size, cap: size}))
+
+			// copy the original char*
+			dst := make([]byte, len(src))
+			copy(dst, src)
+			return *(*string)(unsafe.Pointer(&sliceHeader{buf: &dst[0], len: size, cap: size})), true
```

Comment:
I think you can simplify this. The returned value should be constant, so you can use it directly. Something like this:

```go
	return *(*string)(unsafe.Pointer(&stringHeader{
		ptr: unsafe.Pointer(raw),
		len: size,
	}))
```

Reply:
Also, as you're using a libc anyway it might be easier/simpler to just call `strlen` to get the string length. The current code is hard to read with all the `unsafe` and pointer casts.

Reply:
I am not sure how safe it is actually. Even though we haven't supported `Setenv`, but once we support that, I guess the byte array this string based on would be changed after calling `Setenv` on the same key. What do you think?

Reply:
I see, you are right. From the [Open Group](https://pubs.opengroup.org/onlinepubs/009696799/functions/getenv.html):

> The return value from getenv() may point to static data which may be overwritten by subsequent calls to getenv(), setenv(), or unsetenv().

Reply:
But in that case, you can convert it to a `[]byte` (as you're already doing) and then simply convert it to a `string`. This conversion makes a copy because if the `[]byte` is modified, the string must not be modified. Therefore, you can simplify this to:

```go
src := *(*[]byte)(unsafe.Pointer(&sliceHeader{buf: raw, len: size, cap: size}))
return string(str)
```

Reply:
OK, that looks much simpler. Thanks!